File: common/src/main/java/de/hype/bbsentials/shared/objects/ComGoalPositionExtendedData.java
package de.hype.bbsentials.shared.objects;

import de.hype.bbsentials.shared.packets.function.PositionCommunityFeedback;

import java.time.Instant;
import java.time.temporal.ChronoUnit;

public class ComGoalPositionExtendedData {
    public final Instant fromTime;
    public final String mcuuid;
    public final PositionCommunityFeedback.ComGoalPosition position;

    public ComGoalPositionExtendedData(PositionCommunityFeedback.ComGoalPosition data, String mcuuid, Instant time) {
        this.mcuuid = mcuuid;
        position = data;
        fromTime = time;
    }

    public Instant getReferenceTime() {
        int punPos = (100 - position.position);
        return fromTime.plus((long) punPos * punPos, ChronoUnit.MINUTES);
    }

    public boolean dataEquals(ComGoalPositionExtendedData that) {
        return mcuuid.equals(that.mcuuid) && position.dataEquals(that.position);
    }
}

File: common/src/main/kotlin/de/hype/bingonet/shared/objects/ComGoalPositionExtendedData.kt
package de.hype.bingonet.shared.objects

import de.hype.bingonet.shared.packets.function.PositionCommunityFeedback.ComGoalPosition
import java.time.Instant
import java.time.temporal.ChronoUnit

class ComGoalPositionExtendedData(
    @JvmField val position: ComGoalPosition,
    val mcuuid: String,
    @JvmField val fromTime: Instant
) {
    val referenceTime: Instant?
        get() {
            val position = position.position ?: return null
            val punPos = (100 - position)
            return fromTime.plus(punPos.toLong() * punPos, ChronoUnit.MINUTES)
        }

    fun dataEquals(that: ComGoalPositionExtendedData): Boolean {
        return mcuuid == that.mcuuid && position.dataEquals(that.position)
    }
}
